buildSrc/src/main/kotlin/io/ktor/plugins/registry/KtorReleases.kt

```diff
@@ -42,41 +44,69 @@ private fun writeToFile(versionsList: List<String>) =
             it.parent.createDirectories()
     }.writeText(versionsList.joinToString("\n"))
 
-private fun fetchKtorVersionsFromMaven(): List<String> {
-    val allVersions = buildList {
-        URL(KTOR_MAVEN_REPO).openStream().use {
-            val dbf: DocumentBuilderFactory = DocumentBuilderFactory.newInstance()
-            val db = dbf.newDocumentBuilder()
-            val doc = db.parse(it)
-
-            val versioning = doc.getElementsByTagName("versioning").item(0)
-            val versions = versioning.childNodes
-
-            for (i in 0 until (versions.length)) {
-                if (versions.item(i).nodeName == "versions") {
-                    val versionList = versions.item(i).childNodes
-                    for (j in 0 until versionList.length) {
-                        if (versionList.item(j).nodeName == "version") {
-                            add(versionList.item(j).textContent)
-                        }
-                    }
-                }
+internal fun fetchKtorVersionsFromMaven(
+    latestCount: Int,
+    source: Document = fetchAndParseXml()
+): List<String> {
+    val groupedVersions = source.readVersionNumbers().groupByVersions()
+
+    return groupedVersions.filterByLatest(latestCount)
+        .map(VersionNumber::number)
+}
+
+private fun fetchAndParseXml(location: String = KTOR_MAVEN_REPO): Document {
+    val doc = URL(location).openStream().use { input ->
+        val dbf: DocumentBuilderFactory = DocumentBuilderFactory.newInstance()
+        val db = dbf.newDocumentBuilder()
+        db.parse(input)
+    }
+    return doc
+}
+
+private fun Document.readVersionNumbers(): Sequence<VersionNumber> = sequence {
+    val versioning = getElementsByTagName("versioning").item(0)
+
+    versioning.childNodes("versions").forEach { versionsList ->
+        versionsList.childNodes("version").forEach { version ->
+            try {
+                yield(VersionNumber(version.textContent))
+            } catch (e: Exception) {
+                e.printStackTrace()
```

Comment:
Should we use logging instead?

Reply:
Hmm yeah I'll hook it up with the build log.